File: service/mrdocument/rtf_convert.py
"""RTF to PDF conversion via unrtf (RTF → HTML) and WeasyPrint (HTML → PDF)."""

import logging
import subprocess

from .html_convert import html_to_pdf

logger = logging.getLogger(__name__)


def rtf_to_pdf(rtf_bytes: bytes, filename: str = "document.rtf") -> bytes:
    """
    Convert an RTF file to PDF.

    Uses unrtf to convert RTF → HTML, then WeasyPrint to render HTML → PDF.

    Args:
        rtf_bytes: Raw bytes of the RTF file
        filename: Original filename (for logging)

    Returns:
        PDF file as bytes

    Raises:
        ValueError: If RTF cannot be converted
    """
    logger.debug("Converting RTF to HTML: %s (%d bytes)", filename, len(rtf_bytes))

    try:
        result = subprocess.run(
            ["unrtf", "--html"],
            input=rtf_bytes,
            capture_output=True,
            timeout=30,
        )
    except FileNotFoundError:
        raise ValueError("unrtf is not installed")
    except subprocess.TimeoutExpired:
        raise ValueError(f"RTF conversion timed out for {filename}")

    if result.returncode != 0:
        stderr = result.stderr.decode("utf-8", errors="replace").strip()
        raise ValueError(f"unrtf failed for {filename}: {stderr}")

    html_bytes = result.stdout
    if not html_bytes:
        raise ValueError(f"unrtf produced no output for {filename}")

    logger.info("Converted RTF to HTML: %s (%d -> %d bytes)", filename, len(rtf_bytes), len(html_bytes))
    return html_to_pdf(html_bytes, filename)

File: service-rs/src/rtf_convert.rs
use std::process::Command;
use tracing::{debug, info};

use crate::html_convert::html_to_pdf;

/// Convert RTF bytes to PDF via unrtf (RTF -> HTML) then HTML -> PDF.
pub fn rtf_to_pdf(rtf_bytes: &[u8], filename: &str) -> Result<Vec<u8>, String> {
    debug!(
        "Converting RTF to HTML: {} ({} bytes)",
        filename,
        rtf_bytes.len()
    );

    let output = Command::new("unrtf")
        .arg("--html")
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
        .map_err(|e| {
            if e.kind() == std::io::ErrorKind::NotFound {
                "unrtf is not installed".to_string()
            } else {
                format!("Failed to start unrtf: {}", e)
            }
        })
        .and_then(|mut child| {
            use std::io::Write;
            if let Some(ref mut stdin) = child.stdin {
                stdin
                    .write_all(rtf_bytes)
                    .map_err(|e| format!("Failed to write to unrtf stdin: {}", e))?;
            }
            drop(child.stdin.take());
            child
                .wait_with_output()
                .map_err(|e| format!("unrtf failed: {}", e))
        })?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("unrtf failed for {}: {}", filename, stderr.trim()));
    }

    let html_bytes = output.stdout;
    if html_bytes.is_empty() {
        return Err(format!("unrtf produced no output for {}", filename));
    }

    info!(
        "Converted RTF to HTML: {} ({} -> {} bytes)",
        filename,
        rtf_bytes.len(),
        html_bytes.len()
    );

    html_to_pdf(&html_bytes, filename)
}
